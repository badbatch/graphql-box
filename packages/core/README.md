# @graphql-box/core

The GraphQL Box core module.

[![npm version](https://badge.fury.io/js/%40graphql-box%2Fcore.svg)](https://badge.fury.io/js/%40graphql-box%2Fcore)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)

## Installation

```bash
yarn add @graphql-box/core
```

## Usage

Read the GraphQL Box [usage guide](../../README.md#usage).

## Documentation

Read the module API [documentation](docs/README.md).
