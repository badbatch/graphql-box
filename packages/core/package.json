{
  "name": "@graphql-box/core",
  "version": "4.1.6",
  "description": "The GraphQL Box core module.",
  "author": "Dylan Aubrey <dylanaubrey@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/badbatch/graphql-box",
  "bugs": {
    "url": "https://github.com/badbatch/graphql-box/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/badbatch/graphql-box",
    "directory": "packages/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "main": "lib/main/index.js",
  "module": "lib/module/index.mjs",
  "browser": "lib/browser/index.js",
  "types": "lib/types/index.d.ts",
  "keywords": [
    "graphql-box",
    "graphql",
    "client",
    "server",
    "isomorphic"
  ],
  "scripts": {
    "compile": "concurrently npm:compile:*",
    "compile-watch": "concurrently npm:compile-watch:*",
    "compile-watch:main": "yarn run compile:main --watch",
    "compile-watch:types": "yarn run compile:types --watch",
    "compile:browser": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js",
    "compile:main": "cross-env BABEL_ENV=main babel ./src --out-dir ./lib/main --extensions \".ts\" --source-maps --config-file ../../babel.config.js",
    "compile:module": "cross-env BABEL_ENV=module babel ./src --out-dir ./lib/module --extensions \".ts\" --out-file-extension \".mjs\" --source-maps --config-file ../../babel.config.js",
    "compile:types": "tsc --build",
    "watch": "cross-env BABEL_ENV=browser rollup -c ../../rollup.config.js -w"
  },
  "dependencies": {
    "serialize-error": "8.1.0"
  },
  "peerDependencies": {
    "@babel/runtime": "< 8",
    "core-js": "< 4",
    "lodash": "< 5"
  }
}