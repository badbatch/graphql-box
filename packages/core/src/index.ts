import * as coreDefs from "./defs";
import Core from "./main";

export {
  Core as default,
  coreDefs,
};
