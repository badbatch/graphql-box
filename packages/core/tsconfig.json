{
  "extends": "../tsconfig.settings.json",
  "compilerOptions": {
    "outDir": "lib/types",
    "rootDir": "src"
  },
  "include": [
    "src"
  ]
}