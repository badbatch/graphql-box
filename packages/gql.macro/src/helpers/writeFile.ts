import { appendFileSync } from "fs";

export default (requestWhitelistPath: string, whitelist: string[]) => {
  whitelist.forEach(entry => {
    appendFileSync(requestWhitelistPath, entry);
  });
};
