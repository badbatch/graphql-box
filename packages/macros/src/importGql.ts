import { readFileSync } from 'node:fs';
import { isAbsolute, parse, resolve } from 'node:path';
import { resolveImports } from '#helpers/resolveImports.ts';

export const importGql = (gqlPath: string): string => {
  const fullGqlPath = isAbsolute(gqlPath) ? gqlPath : resolve(process.cwd(), gqlPath);
  const { dir } = parse(fullGqlPath);
  return resolveImports(readFileSync(fullGqlPath, { encoding: 'utf8' }), dir);
};
