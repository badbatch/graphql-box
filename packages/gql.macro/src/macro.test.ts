import { transformSync } from '@babel/core';
import { expect, jest } from '@jest/globals';
import { writeFile } from './helpers/writeFile.ts';

jest.mock('./helpers/writeFile.ts', () => ({
  writeFile: jest.fn(),
}));

const transformSourceFile = (sourceFile: string, macroConfig: Record<string, unknown> = {}) => {
  const transformed = transformSync(sourceFile, {
    filename: 'test.js',
    plugins: [
      [
        'macros',
        {
          gql: {
            ...macroConfig,
            // eslint-disable-next-line unicorn/prefer-module
            basePath: __dirname,
          },
        },
      ],
    ],
  })?.code;

  return transformed;
};

describe('gql.macro', () => {
  describe('when the macro is a tag call', () => {
    let transformed: string | null | undefined;

    beforeAll(() => {
      const sourceFile = `
      import importGql from "./packages/gql.macro/src/macro";

      const GET_CONFIG = importGql\`./__testUtils__/gql/queries/GetConfig.gql\`;
      export default GET_CONFIG;
    `;

      transformed = transformSourceFile(sourceFile);
    });

    it('should transform the macro reference correctly', () => {
      expect(transformed).toMatchSnapshot();
    });

    it('should call writeFile the correct number of times', () => {
      expect(writeFile).toHaveBeenCalledTimes(1);
    });

    it('should call writeFile with the correct arguments', () => {
      expect(writeFile).toHaveBeenCalledWith(`${process.cwd()}/requestWhitelist.txt`, [
        '33771f995332f7d2c4f83258a37d3108',
      ]);
    });
  });
});
