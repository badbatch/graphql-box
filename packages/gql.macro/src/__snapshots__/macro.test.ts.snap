// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gql.macro when the macro is a tag call should transform the macro reference correctly 1`] = `
""use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
const GET_CONFIG = "\\nquery GetConfig {\\n  config {\\n    ...ConfigImages\\n  }\\n}\\n\\n\\nfragment ConfigImages on Config {\\n  images {\\n    backdropSizes\\n    logoSizes\\n    posterSizes\\n    profileSizes\\n    stillSizes\\n    ...ImageUrls\\n  }\\n}\\n\\nfragment ImageUrls on Image {\\n  baseUrl\\n  secureBaseUrl\\n}\\n";
var _default = exports.default = GET_CONFIG;"
`;
